### ISSUE TITLE
(stepfunction): JsonPath does not allow intrinsic functions

### ISSUE THREAD

**Issue by sannies** [aws/aws-cdk#15287]
Currently a JSON path values must be exactly '$', '$$', start with '$.', start with '$$.' or start with '$['. 
But this is not correct as Parameters allow [intrinsic functions](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html) and can thereby also start with 'States.Format', 'States.StringToJson', 'States.JsonToString' or 'States.Array'. 

### Reproduction Steps

```
import * as cdk from '@aws-cdk/core';
import {Duration} from "@aws-cdk/core";
import {StateMachine, Pass, JsonPath} from "@aws-cdk/aws-stepfunctions";


export class JsonPathStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const definition = new Pass(scope, 'pass',{
      parameters: {
        'inputAsString' : JsonPath.stringAt('States.JsonToString($.input)')
      }
    })

    new StateMachine(this, 'StateMachine', {
      definition,
    });
  }
}

```

### What did you expect to happen?

The stack should synthesise the step like this:

```
"Pass": {
      "Type": "Pass",
      "End": true,
      "Parameters": {
        "inputAsString.$": "States.JsonToString($.input)"
      }
    }
```

### What actually happened?

cdk deploy fails:

```
cdk deploy
/Users/sannies/dev/JsonPath/node_modules/@aws-cdk/aws-stepfunctions/lib/fields.ts:139
    throw new Error(`JSON path values must be exactly '$', '$$', start with '$.', start with '$$.' or start with '$[' Received: ${path}`);
          ^
Error: JSON path values must be exactly '$', '$$', start with '$.', start with '$$.' or start with '$[' Received: States.JsonToString($.
    at validateJsonPath (/Users/sannies/dev/JsonPath/node_modules/@aws-cdk/aws-stepfunctions/lib/fields.ts:139:11)
    at Function.stringAt (/Users/sannies/dev/JsonPath/node_modules/@aws-cdk/aws-stepfunctions/lib/fields.ts:11:5)
    at new JsonPathStack (/Users/sannies/dev/JsonPath/lib/json_path-stack.ts:12:36)
    at Object.<anonymous> (/Users/sannies/dev/JsonPath/bin/json_path.ts:7:1)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Module.m._compile (/Users/sannies/dev/JsonPath/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/sannies/dev/JsonPath/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)

```

### Environment

  - **CDK CLI Version  :** 1.108.1
  - **Framework Version:** 1.108.1
  - **Node.js Version:** v16.1.0
  - **OS               :** macOS Catalina 10.15.7 
  - **Language (Version):** TypeScript (3.8.3) but will obviously also happen in all others

### Other

I'd propose to ease the validation in https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-stepfunctions/lib/fields.ts#L216

from

```
function validateJsonPath(path: string) {
  if (path !== '$'
    && !path.startsWith('$.')
    && path !== '$$'
    && !path.startsWith('$$.')
    && !path.startsWith('$[')
  ) {
    throw new Error(`JSON path values must be exactly '$', '$$', start with '$.', start with '$$.' or start with '$[' Received: ${path}`);
  }
}

```

to 

```
function validateJsonPath(path: string) {
  if (path !== '$'
    && !path.startsWith('$.')
    && path !== '$$'
    && !path.startsWith('$$.')
    && !path.startsWith('$[')
    && !path.startsWith('States.')
 
 ) {
    throw new Error(`JSON path values must be exactly '$', '$$', start with '$.', start with '$$.' or start with '$[' Received: ${path}`);
  }
}
```

--- 

This is :bug: Bug Report

**Comment by peterwoodworth** [comment]
Thank you for submitting this issue, and providing the easily reproducible code, I can't find a workaround for this using the CDK. I'm not sure if CloudFormation supports this, however my gut tells me that it does and this is just a CDK issue. Here's where the CDK is throwing the error https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-stepfunctions/lib/fields.ts#L216-L225
Some adjustments to this function are all the CDK should need to support this, if the case is that CFN will allow for this type of input

**Comment by sannies** [comment]
For the sake of completeness and may it help others: There is an ugly workaround. By overriding the steps class - in my case DynamoUpdateItem - I can manipulate the steps JSON in its dict representation like that:

```
        class MyDynamoUpdateItem(DynamoUpdateItem):

            def to_state_json(self) -> typing.Mapping[typing.Any, typing.Any]:
                state_json = super().to_state_json()
                dataAttr = state_json['Parameters']['ExpressionAttributeValues'][':data']['S']
                del state_json['Parameters']['ExpressionAttributeValues'][':data']['S']
                state_json['Parameters']['ExpressionAttributeValues'][':data']['S.$'] = dataAttr
                return state_json
```

**Comment by michaelwiles** [comment]
See #11286

**Comment by stenzing** [comment]
Just a comment. 

```python
sfn_tasks \
                .LambdaInvoke(self, aName,
                              lambda_function=aLambda,
                              payload=sfn.TaskInput.from_object({
                                  "org-id": sfn.JsonPath.string_at("$.ORG_ID"),
                                  "properties.$":"States.StringToJson($.PROPERTIES)"
                              }),
                              timeout=Duration.seconds(300))
```

is working as expected. Not even using that JsonPath object, the only thing to note to add the .$ on the end of the key.

**Comment by jihndai** [comment]
This may have been fixed by #15320

**Comment by sannies** [comment]
That was resolved at some point. Works in 2.8 at least.
